File: JavaScript/40-Combination Sum II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort(function(a,b){return a-b;});
    // console.log(candidates)
    let res = [];
    
    let dfs = (start, val, remain) =>{
        if (remain < 0) return;
        if (remain === 0) res.push(val);
        for (let i=start; i < candidates.length; i++){
            if (i > start && candidates[i] === candidates[i-1]) continue;
            if (remain < candidates[i]) break;
            dfs(i+1, val.concat([candidates[i]]), remain-candidates[i]);
        }
    }
    dfs(0, [], target);
    return res;
};

File: C++/40-Combination Sum II.cpp
class Solution {
public:
    int N;
    void dfs(int start, vector<int>& curr, vector<vector<int>>& res, vector<int>& candidates, int remain) {
        if (remain == 0) {
            res.push_back(curr);
            return;
        }
        for (int i = start; i<N; i++) {
            if (remain < candidates[i]) break;
            if (i == start || candidates[i] != candidates[i-1]) { // pruning, avoid duplicate cases
                curr.push_back(candidates[i]);
                // dfs(i+1, curr + {candidates[i]}, res, candidates, remain-candidates[i]);   // cannot do this
                dfs(i+1, curr, res, candidates, remain-candidates[i]);
                curr.pop_back();
            }
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        this->N = candidates.size();
        sort(candidates.begin(), candidates.end()); // sort for pruning
        vector<vector<int>> res;
        vector<int> curr;
        dfs(0, curr, res, candidates, target);

        return res;
    }
};